datafusion/proto/src/from_proto.rs

```diff
@@ -838,6 +754,23 @@ impl TryFrom<&protobuf::ScalarValue> for ScalarValue {
     }
 }
 
+/// Ensures that all `values` are of type DataType::List and have the
+/// same type as field
+fn validate_list_values(field: &Field, values: &[ScalarValue]) -> Result<(), Error> {
```

Comment:
I am not sure how valuable it is to do this validation if serialized `ScalarValue` always came from DataFusion, but I wanted to keep the same semantics

Reply:
It would be a weird edge case but if it did happen it would be hard to debug from whatever error happened downstream so I think it's worthwhile to do here.